src/piadm.sh

```diff
@@ -0,0 +1,741 @@
+	mapfile -t pis \
+		< <(cd /$bootfs ; cat platform-*/etc/version/platform)
+	for pi in ${pis[@]}; do
+	    if [[ $activestamp == $pi ]]; then
+		active="yes"
+	    else
+		active="no"
+	    fi
+	    if [[ $bootstamp == $pi ]]; then
+		booting="yes"
+	    else
+		booting="no"
+	    fi
+	    if [[ $bootbitsstamp == $pi ]]; then
+		bootbits="next"
+	    elif [[ -d boot-$pi ]]; then
+		bootbits="available"
+	    else
+		bootbits="none"
+	    fi
+	    printf "%-18s %-30s %-12s %-5s %-5s\n" $pi $bootfs $bootbits \
+		$active $booting
+	done
+    done
+}
+
+update_boot_sectors() {
+    pool=$1
+    bootfs=$2
+
+    # XXX WARNING -- illumos#12894 will allow slogs.  We will need to
+    # alter the generation of boot_devices accordingly.
+    # Generate the pool's boot devices now, in case we did something
+    # hyper-clever for the pool.  s1 may be created, but not yet PCFS...
+    mapfile -t boot_devices < <(zpool list -v "$pool" | \
+        grep -E 'c[0-9]+' | awk '{print $1}' | sed -E 's/s[0-9]+//g')
+
+    # Reality check the pool was created with -B.
+    # First way to do this is to check for the `bootsize` property not
+    # its default, which is NO bootsize.
+    zpool get bootsize $pool | grep -q -w default
+    if [[ $? -eq 0 ]]; then
+	# No bootsize is a first-cut test.  It passes if the pool was
+	# created with `zpool create -B`. There's one other that needs
+	# to be peformed, because some bootable pools are manually
+	# configured to share slices with other functions (slog,
+	# l2arc, dedup):
+
+	# Use fstyp to confirm if this is a stealth EFI booter...
+	type=$(fstyp /dev/dsk/${boot_devices[0]}s0)
+	if [[ "$type" == "pcfs" ]]; then
+	    # If we detect PCFS on s0, it's LIKELY an EFI System Partition that
+	    # was crafted manually.  Use s1 if it's ZFS, or bail if it's not.
+
+	    s1type=$(fstyp /dev/dsk/${boot_devices[0]}s1)
+	    if [[ "$s1type" != "zfs" ]]; then
+		fatal "Unusual configuration, ${boot_devices[0]}s1 not ZFS"
+	    fi
+	    suffix=s1
+	else
+	    suffix=s0
+	fi
+    else
+	# Guaranteed that s0 is EFI System Partition, ZFS lives on s1.
+	suffix=s1
+    fi
+
+    some=0
+    for a in "${boot_devices[@]}"; do
+	# Plow through devices, even if some fail.  installboot also
+	# does loader-into-EFI-System-Partition this way.
+	# Trailing / is important in the -b argument because boot is actually
+	# a symlink.
+	installboot -m -b /${bootfs}/boot/ /${bootfs}/boot/pmbr \
+	    /${bootfs}/boot/gptzfsboot \
+	    /dev/rdsk/${a}${suffix} > /dev/null 2>&1 || \
+	    echo "WARNING: Can't installboot on ${a}${suffix}"
+	if [[ $? -eq 0 ]]; then
+	    some=1
+	fi
+    done
+    if [[ $some -eq 0 ]]; then
+	fatal "Could not installboot(1M) on ANY vdevs of pool $2"
+    fi
+
+}
+
+activate() {
+    pistamp=$1
+    piname_present_get_bootfs $pistamp $2
+    pool=$(echo $bootfs | awk -F/ '{print $1}')
+
+    cd /$bootfs
+    bootstamp=$(file -h platform | awk '{print $5}' | sed 's/\.\/platform-//g')
```

Comment:
readlink means you're not parsing english text

Reply:
Hmmm, I can use an entry in that directory, can't I?  `bootstamp=$(cat platform/etc/version/platform)`  Will fix.